### DISCUSSION TITLE
Does Renovate have an incorrect implementation of "Minor" versions according to the  semver spec?

### DISCUSSION THREAD

**Discussion by Mknight492** [renovatebot/renovate#21637] [Request Help] [answered]
### Type of discussion.

I'm proposing an idea

### Tell us more.

I've been using renovate at work and have come across an interesting case.

We're batching all minor and patch version updates into a single PR to reduce the overhead or reviewing the PRs.

Using something like so:

``` json
  "packageRules": [
    {
      "matchManagers": "nuget",
      "groupName": "All non-major nuget dependencies",
      "matchUpdateTypes": ["minor", "patch"],
      "excludePackageNames": ["dotnet-sdk"]
    },
    ]
```

However in this case packages with a Major version zero (0.y.z) are getting included in the PRs as well. I wonder if these packages should be excluded from "minor" and "patch" version changes as Major version zero packages have unstable APIs according to the [semver spec](https://semver.org/#spec-item-4). The whole point of semver is about the stability of the APIs NOT necessarily if the changes is breaking or not.

There was a [similar discussion here](https://github.com/renovatebot/renovate/discussions/15093). However my understanding was that Patch, Minor and Major versioning doesn't apply to a 0.y.z releases and I can't see how it makes any sense for  "Minor" update to apply to them.

If you read the semver documentation it state that:
 "Version 1.0.0 defines the public API. The way in which the version number is incremented **after** this release is dependent on this public API and how it changes." 
 
 and: 
 "Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the **public API**"
 
So **if the "public API" doesn't exist according to the spec how can it it be a Minor Version update**?

Edit:
I suspect I'm just misreading the semver docs but I wonder if it would still be useful adding some comment about this in the docs? something along the line of if you're going to rely on the "minor" and "patch" version in order to reduce the risk of breaking changes you should also consider excluding Major version zero  packages (0.y.z). This can be achieved by..."

**Comment by rarkins** [comment, marked as answer]
Renovate's implementation is correct. Exact same misunderstanding and answer as #15093

**Reply by Mknight492** [reply]
Sorry to bother you then.

Is it worth adding anything to the docs to warn users about this?

> I suspect I'm just misreading the semver docs but I wonder if it would still be useful adding some comment about this in the docs? something along the line of if you're going to rely on the "minor" and "patch" version in order to reduce the risk of breaking changes you should also consider excluding Major version zero packages (0.y.z). This can be achieved by..."

**Reply by HonkingGoose** [reply]
I'm adding a warning to the `matchUpdateTypes` docs:

- https://github.com/renovatebot/renovate/pull/21747

We also have a warning here:

https://docs.renovatebot.com/key-concepts/automerge/#automerge-non-major-updates

---

To fix your problem, add this `matchCurrentVersion` rule to your config:

```json
  "packageRules": [
    {
      "matchManagers": "nuget",
      "groupName": "All non-major nuget dependencies",
      "matchUpdateTypes": ["minor", "patch"],
      "excludePackageNames": ["dotnet-sdk"],
      "matchCurrentVersion": "!/^0/"
    },
    ]
```
